README
				 Gnome2::Wnck
				--------------

This module allows you to use the Window Navigator Construction Kit library
(libwnck for short) from Perl.

The index of the automatically generated API documentation can be accessed
with:

  perldoc Gnome2::Wnck::index

				 Installation
				--------------

To install Gnome2::Wnck run the following commands:

  perl Makefile.PL
  make
  make test
  make install

				 Dependencies
				--------------

Gnome2::Wnck depends on:

  ExtUtils::Depends >= 0.20
  ExtUtils::PkgConfig >= 1.03
  Glib >= 1.04
  Gtk2 >= 1.00
  libwnck >= 2.20.0

If you need support for libwnck < 2.20.0, use Gnome2::Wnck 0.14.


NEWS
Overview of changes in Gnome2::Wnck 0.15
========================================

* Require libwnck >= 2.20.0 due to API changes that can't be worked around.
* Adapt to API changes:
  - wnck_pager_set_orientation now returns a boolean
  - wnck_selector_new doesn't take a screen parameter anymore
* Wrap new API:
  - wnck_screen_get_workspaces, wnck_screen_get_window_manager_name, and
    wnck_screen_get_number
  - wnck_window_get_client_window_geometry, wnck_window_set_sort_order,
    wnck_window_is_below, wnck_window_make_below, and wnck_window_unmake_below
  - wnck_workspace_get_screen, wnck_workspace_get_layout_row,
    wnck_workspace_get_layout_column, and wnck_workspace_get_neighbor

Overview of changes in Gnome2::Wnck 0.14
========================================

* Add new binding for wnck_window_set_geometry.

Overview of changes in Gnome2::Wnck 0.13
========================================

* Add support for WnckSelector.
* Don't run the pager test by default to avoid changing the user's current
  pager settings.

Overview of changes in Gnome2::Wnck 0.12
========================================

* Add Gnome2::Wnck::Window::is_above, make_above, and unmake_above.

Overview of changes in Gnome2::Wnck 0.11
========================================

* Adapt to upstream API changes.  Require libwnck 2.12.0.
* Change some internal details.

Overview of changes in Gnome2::Wnck 0.10
========================================

* Fix some version checks.
* Accomodate for upstream API changes.
* Add new API:
  - wnck_window_get_sort_order,
  - wnck_window_or_transient_demands_attention, and
  - wnck_window_transient_is_active.

Overview of changes in Gnome2::Wnck 0.09
========================================

* Fix test failure.

Overview of changes in Gnome2::Wnck 0.08
========================================

* New bindings for:
  - wnck_screen_get_previously_active_window
  - wnck_window_is_fullscreen
  - wnck_window_set_fullscreen
  - wnck_window_is_most_recently_activated

Overview of changes in Gnome2::Wnck 0.07
========================================

* Bind and test wnck_window_get_window_type, wnck_window_demands_attention,
  wnck_tasklist_set_grouping, and wnck_pager_set_display_mode.
* Remove the binding for wnck_gtk_window_set_dock_type.

Overview of changes in Gnome2::Wnck 0.06
========================================

* Make we sure we compile against libwnck 2.6.2.1.

Overview of changes in Gnome2::Wnck 0.05
========================================

* Improve compatibility with older versions of libwnck.
* Make sure we work with libwnck 2.6.3, which introduces an incompatible API
  change -- in a stable release.


Wnck.pm
package Gnome2::Wnck;

# $Header: /cvsroot/gtk2-perl/gtk2-perl-xs/Gnome2-Wnck/Wnck.pm,v 1.21 2007/12/09 14:09:57 kaffeetisch Exp $

use 5.008;
use strict;
use warnings;

use Gtk2;

require DynaLoader;

our @ISA = qw(DynaLoader);

our $VERSION = '0.15';

sub import {
  my $self = shift();
  $self -> VERSION(@_);
}

sub dl_load_flags { $^O eq 'darwin' ? 0x00 : 0x01 }

Gnome2::Wnck -> bootstrap($VERSION);

1;
__END__

=head1 NAME

Gnome2::Wnck - Perl interface to the Window Navigator Construction Kit

=head1 SYNOPSIS

  use Gnome2::Wnck;

  my $screen = Gnome2::Wnck::Screen -> get_default();
  $screen -> force_update();

  my $pager = Gnome2::Wnck::Pager -> new($screen);
  my $tasklist = Gnome2::Wnck::Tasklist -> new($screen);

=head1 ABSTRACT

This module allows a Perl developer to use the Window Navigator Construction
Kit library (libwnck for short) to write tasklists and pagers.

=head1 SEE ALSO

L<Gnome2::Wnck::index>(3pm), L<Gtk2>(3pm), L<Gtk2::api>(3pm) and the source
code of libwnck.

=head1 AUTHOR

Torsten Schoenfeld E<lt>kaffeetisch@web.deE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2006 by the gtk2-perl team

=cut


wnck2perl.h
/*
 * Copyright (C) 2003 by the gtk2-perl team (see the file AUTHORS)
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Header: /cvsroot/gtk2-perl/gtk2-perl-xs/Gnome2-Wnck/wnck2perl.h,v 1.6 2005/09/17 20:30:37 kaffeetisch Exp $
 */

#ifndef _WNCK2PERL_H_
#define _WNCK2PERL_H_

#include <gtk2perl.h>

#include <libwnck/libwnck.h>
#include <libwnck/window-action-menu.h>

#include "wnck2perl-version.h"

#if !WNCK_CHECK_VERSION (2, 10, 0)
# include <libwnck/window-menu.h>
#endif

#if WNCK_CHECK_VERSION (2, 6, 0)
# include <libwnck/class-group.h>
#endif

#include "wnck2perl-autogen.h"

#endif /* _WNCK2PERL_H_ */


Makefile.PL
#!/usr/bin/perl

BEGIN { require 5.008; }

use strict;
use warnings;

use ExtUtils::MakeMaker;
use Cwd;
use File::Spec;

# $Header: /cvsroot/gtk2-perl/gtk2-perl-xs/Gnome2-Wnck/Makefile.PL,v 1.16 2007/12/09 14:09:56 kaffeetisch Exp $

my %build_reqs = (
  'perl-ExtUtils-Depends'   => '0.20',
  'perl-ExtUtils-PkgConfig' => '1.03',
  'perl-Glib'               => '1.04',
  'perl-Gtk2'               => '1.00',
  'Wnck'                    => '2.20.0',
);

my %pm_reqs = (
  'Glib'                => $build_reqs{'perl-Glib'},
  'Gtk2'                => $build_reqs{'perl-Gtk2'},
  'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
  'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
);

unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Gtk2 '$build_reqs{'perl-Gtk2'}';"
           . "use Glib::MakeHelper;"
           . "use Gtk2::CodeGen;"
           . "1") {
  warn "$@\n";
  WriteMakefile(
    PREREQ_FATAL => 1,
    PREREQ_PM    => \%pm_reqs,
  );
  exit 1; # not reached
}

# --------------------------------------------------------------------------- #

my %pkgcfg = ExtUtils::PkgConfig->find("libwnck-1.0 >= $build_reqs{Wnck}");

mkdir 'build', 0777;

my @xs_files = <xs/*.xs>;
my %pod_files = (
  'Wnck.pm' => '$(INST_MAN3DIR)/Gnome2::Wnck.$(MAN3EXT)',
  Glib::MakeHelper->do_pod_files (@xs_files),
);

Gtk2::CodeGen->parse_maps('wnck2perl');
Gtk2::CodeGen->write_boot(ignore => qr/^Gnome2::Wnck$/);

ExtUtils::PkgConfig->write_version_macros (
  "build/wnck2perl-version.h",
  "libwnck-1.0" => "WNCK",
);

# --------------------------------------------------------------------------- #

my $wnck = ExtUtils::Depends->new('Gnome2::Wnck', 'Gtk2');
$wnck->set_inc($pkgcfg{cflags} . " -DWNCK_I_KNOW_THIS_IS_UNSTABLE");
$wnck->set_libs($pkgcfg{libs});
$wnck->add_xs(@xs_files);
$wnck->add_pm('Wnck.pm' => '$(INST_LIBDIR)/Wnck.pm');
$wnck->add_typemaps(map {File::Spec->catfile(cwd(), $_)} 'build/wnck2perl.typemap');

$wnck->install(qw(wnck2perl.h build/wnck2perl-autogen.h build/wnck2perl-version.h));
$wnck->save_config('build/IFiles.pm');

WriteMakefile(
  NAME          => 'Gnome2::Wnck',
  VERSION_FROM  => 'Wnck.pm', # finds $VERSION
  ABSTRACT_FROM => 'Wnck.pm', # retrieve abstract from module
  XSPROTOARG    => '-noprototypes',
  MAN3PODS      => \%pod_files,
  PREREQ_PM     => \%pm_reqs,
  $wnck->get_makefile_vars,
);

sub MY::postamble {
  return Glib::MakeHelper->postamble_clean ()
       . Glib::MakeHelper->postamble_docs_full (
           DEPENDS => $wnck,
           COPYRIGHT_FROM => 'copyright.pod')
       . Glib::MakeHelper->postamble_rpms (
           GNOME_WNCK => $build_reqs{'Wnck'},
           PERL_EXTUTILS_DEPENDS => $build_reqs{'perl-ExtUtils-Depends'},
           PERL_EXTUTILS_PKGCONFIG => $build_reqs{'perl-ExtUtils-PkgConfig'},
           PERL_GLIB => $build_reqs{'perl-Glib'},
           PERL_GTK => $build_reqs{'perl-Gtk2'});
}
